### ISSUE TITLE
[1.4.x] Windows package installation ETIMEDOUT error

### ISSUE THREAD

**Issue by tcastelli** [meteor/meteor#8013]
I have been experiencing this problem and as discussed on #6960, It seems this could be related to Windows trying to dowload a lot of packages at the same time(every time this happened to me it was the case). I tried to change METEOR_WAREHOUSE_ULRBASE to cloudfront but it was still failing so it doesn't seem something related to the CDN.

Possible fixes could go from trying to download in groups of 2 as @abernix suggested or to have a pool of WAREHOUSE_URLBASES and try different alternatives automatically (this could also solve possible problems for non Windows users).

Reproducing this problem is not always possible, but when it happens you just need to dowload project with a large (+30 atmoshpere packages) for the first time (not present in the .meteor/packages) folder.

> These are the details of the problem:
> 
> maxcdn
> 
>     You are hitting the MaxCDN Paris Datacenter
> 
> amazonwsip
> 
>     84.76.115.221
> 
> tracert
> 
>     static-meteor.netdna-ssl.com [94.31.29.248]
> 
>     1 2 ms 3 ms 2 ms LIVEBOX [192.168.1.1]
>     2 26 ms 25 ms 25 ms 84.76.115.254
>     3 26 ms 29 ms 25 ms 10.255.205.113
>     4 23 ms 24 ms 24 ms 10.255.204.41
>     5 29 ms 27 ms 27 ms 10.34.212.157
>     6 * * * Tiempo de espera agotado para esta solicitud.
>     7 * * * Tiempo de espera agotado para esta solicitud.
>     8 * * * Tiempo de espera agotado para esta solicitud.
>     9 47 ms 43 ms 43 ms 10.34.34.45
>     10 44 ms 43 ms 43 ms bundle-ether101.madtr3.Madrid.opentransit.net [193.251.249.1]
>     11 43 ms 44 ms 45 ms be3012.ccr22.mad05.atlas.cogentco.com [130.117.14.137]
>     12 55 ms 49 ms 48 ms be2293.ccr22.bio02.atlas.cogentco.com [130.117.50.25]
>     13 61 ms 61 ms 61 ms be2423.ccr42.par01.atlas.cogentco.com [130.117.50.77]
>     14 62 ms 60 ms 61 ms be2108.agr22.par01.atlas.cogentco.com [130.117.50.133]
>     15 61 ms 62 ms 62 ms be2947.rcr12.b019498-0.par01.atlas.cogentco.com [154.25.7.105]
>     16 63 ms 64 ms 64 ms netdna-gw.cdg.ip4.cogentco.com [149.6.114.82]
>     17 63 ms 63 ms 63 ms 94.31.29.248.IPYX-077437-ZYO.above.net [94.31.29.248]
> 
> error
> 
>     Errors prevented startup:
> 
>     While downloading alanning:roles@1.2.15...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:autoform@5.8.1...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:collection2@2.10.0...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:collection2-core@1.2.0...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:schema-deny@1.1.0...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:schema-index@1.1.1...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:simple-schema@1.5.3...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading aldeed:template-extension@4.0.0...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     While downloading amplify@1.0.0...:
>     error: connect ETIMEDOUT 54.192.62.173:443
> 
>     ..etc
>

**Comment by abernix** [comment]
Thank you for opening this separate ticket.  There are a few possibilities here (and maybe more!) and I've actually seen this before:

* **The CDN is actually at fault**: As we debugged [in the other issue](https://github.com/meteor/meteor/issues/6960#issuecomment-258408965) you have basically eliminated the (let's say) previously-problematic CDN and in the example above you're using the alternative CloudFront CDN with the same result.  You could try switching back to the primary CDN by _un_-setting `METEOR_WAREHOUSE_URLBASE` (or making it blank).  Also, it should be pointed out that your traceroute is actually going to the primary CDN, not the same IP that the files are trying to download from (54.192.62.173 = cloudfront).

* **Your Internet is _actually_ failing**: Your traceroute shows pretty fast pings so I'm not of the belief this is actually the problem.

* **Meteor is being too aggressive**: Something believe I observed while reviewing #7637 (though I haven't confirmed) was that Meteor seems to trying to download ALL of its packages at once, in parallel.  While potentially fine, this high level of parallelism might present a bit too much for various levels of networking appliances along the way (frankly, something might flag a high number of concurrent connections as some sort of attack).  Up until ~2014 the HTTP 1.1 RFC stated that there shouldn't be more than 2 TCP connections.  This was removed from the from the RFC in 2014 and browsers haven't respected this for some time: For example, I think Chrome has allowed 6 for years.  Anyhow, I think it's unlikely that any of these package downloads are actually finishing faster in this manner so it might be worth slowing them down.  Maybe Windows doesn't like all these connections?  (Personally, my Windows generally runs in a VirtualBox with almost no memory or CPU though [and thus poor network performance] so I'm not entirely surprised that I have seen this).

* **Something broke**: It's entirely possible that the change that was removed in #7637 somehow prevented this – though I'm not sure why.  Additionally, I'm fairly sure @sebakerckhof uses Windows (at least part time) and was involved in submitting and testing that PR.

What version of Windows are you using?  And what version of Meteor is your app using?

**Comment by tcastelli** [comment]
I tried with both CDNs:
first without changing warehouse env variable (example trace) and later again by using the warehouse env variable with cloudfront url (I forgot to save the trace for that time) and both failed. Between attempts I closed and opened the cmd console again to make sure different warehouses were used (original and cloudfront with set METEOR_WA...).
Maybe both of them were failing at that time, but seems unlikely.

So I would say Meteor being too aggresive or something broke have more chances to be the source of the problem. I'm using windows 10 x64, Meteor 1.4.2 (and 1.4.2.1-rc.0)

**Comment by abernix** [comment]
I will try to debug better the next time it happens to me.

As originally reported, I think this only happens when large numbers of packages are updating concurrently and seems to only happen on Windows (suggesting potential fragility in the Windows networking stack as the Meteor code for this is the same on both platforms).

I believe the work-around is to simply run the `meteor update` a couple (few?) more times and it will work itself out.

**Comment by abernix** [comment]
So this happened to me today on Windows 10 Insider Preview (which I run inside a VirtualBox) – I haven't been able to go back to the official Windows release to see if it's happening there too.

Indeed, each time I ran `meteor update`, I got a little bit further.  By enabling the `METEOR_LOG=debug` and `NODE_DEBUG=net,request` env variables (with `$env:METEOR_LOG="debug"` and `$env:NODE_DEBUG="request,net"` in PowerShell, respectively), I was able to get a glimpse of what was going on.  Practically every HTTP request was having to be retried, though occasionally it would go through.  This was happening on both CDNs and my OS X host was working _generally_ fine, but definitely not perfect.  I am seeing general slowness in my DNS resolution today.

This brought me to [this code](https://github.com/meteor/meteor/blob/b1df3ed0ed2b686e96d89ee66252f447b3c3b87e/tools/utils/http-helpers.js#L310) which happens to just be the call to the [`request`](https://www.npmjs.com/package/request) NPM module.  After adding some debug (`Console.debug("There was an error.", error, response, body);`), I found `request` was error-ing with one `ECONNRESET` after another.

Changing Meteor to use single-concurrency downloading (instead of doing them in parallel) didn't help.

The good news is, I was able to reproduce **without** Meteor at all and on both CDNs using only the [`request`](https://www.npmjs.com/package/request) NPM module.  Using this [relatively basic node repo](https://github.com/abernix/meteor-8013-req-fail) the `ECONNRESET` errors continued to come – aside from an occasional, but very rare, success.  Bizarrely, after using `mitmproxy` on my OS X host I was able to inspect the requests and the (code) `200` response `body` was fine – so fine in fact, that the MD5-sums of the downloaded content matched the ones produced on OS X from `curl`, and also matched the `ETag` headers (which are MD5 sums of the file in the case of both CloudFront and MaxCDN) provided by the host.

The additional good news is that one out of 30 attempts usually went through and once it was downloaded, obviously the local caches continued to work fine and everything was fine again until I blew away my .  Anyhow, long story short, this appears to be either a bug with Windows, VirtualBox, `request` or a combination thereof.  Unfortunately, a combination of these things makes it relatively hard issues to track down known issues about it.

Anyhow, thought I'd document this for future reference.  I'll try to revisit it again soon, but as for now this has been too much of a tangent from what I was actually looking at. :smile:

**Comment by abernix** [comment]
I'm going to close this issue as it really hasn't picked up any other traction from other users.

In my own findings on this, this was only occurring under very specific network conditions.  For me, this meant running Windows in VirtualBox and via the "NAT" networking method and other networking methods (Bridged, etc.) did not demonstrate the same problem.  Some of my dialog about this revelation is included in [this commit](https://github.com/abernix/meteor-8013-req-fail/commit/846e5c89b2db5a9ce0853e1a96456c8f0621b0ed) on the reproduction repository I linked above.

Thanks for reporting this originally!
